### DISCUSSION TITLE
How to add another IoT protocol?

### DISCUSSION THREAD

**Discussion by kryslt** [arendst/Tasmota#20130] [Support request] [answered]
Hello,
is it possible to integrate additional IoT protocols to Tasmota?

We use our own very lightweight IoT protocol between our own devices.
The goal is to integrate directly with already tasmotized devices (after firmware upgrade of course).
I know we can use HTTP or MQTT or maybe others but it is not suitable for us.

I am completely new to Tasmota.

Regards, Tomas

**Comment by sfromis** [comment]
Adding another protocol would entail coding a driver, and quite a bit of work to have it well-integrated. The "M" in the "Tasmota" name is for "MQTT", hence many concepts are designed for this.

With ESP32 devices, it may be "more approachable" coding such a driver in Berry, in a similar way to what is done with Tasmota support for Matter.

More generally, a simpler strategy would probably be not to add a protocol, but instead having a bridge device handling both MQTT and "your own" protocol. This is how Tasmota supports Matter with ESP8266 devices not able to use Matter directly, but going via some ESP32 device with Tasmota and Matter.

**Reply by kryslt** [reply]
Thank you much for answer. Will look at the Berry, never heard of it before. 
We basically need only a single broadcast message (UDP) used for automatic pairing and a single data message (TCP).
The problem with a bridge is that you need another piece of hardware (or software running somewhere), and we generally want to avoid that. For our purposes, MQTT is also a very bad choice.

**Reply by s-hadinger** [reply]
As said by @sfromis if you are using ESP32 or derivative (ESP32C3/S2/S3/S6...) Berry should make this quite easy to do.

**Comment by kryslt** [comment, marked as answer]
I reply to myself.
Because the aforementioned Berry thing is just a Python like scripting language, and we have all our IoT protocol sources in Ansi C, we gave up on the idea of direct integration of our protocol into Tasmota. Instead, we decided to communicate via HTTP commands and make our own simple HTTP bridge (only a small piece of software) which will be incorporated into our firmware.
